### PR TITLE
feat!: Add support for Nuxt 3

### PR DESCRIPTION
**Pull request by dankerow**
This PR enables this module to work for  Nuxt 3.

# Notable changes
- Uses @nuxt/module-builder to build module.
- Refactor structure (lib -> src, templates -> runtime)
- Rewrite in TypeScript
- Move from templating options to runtime config

# TODO
- Enable e2e test suite (https://github.com/nuxt/framework/pull/2952)

### CODE REVIEW

File: playground/pages/index.vue [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  <div>
+    Works!
+  </div>
+</template>
+
+<script setup>
+defineNuxtComponent({
+  name: 'TheIndex'
+})
+
```
Comment (danielroe) [inline playground/pages/index.vue:11]: This won't have any effect within script setup.
```suggestion
```

File: playground/app.vue [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<template>
+	<div>
+		<NuxtLayout>
+      <NuxtPage />
+		</NuxtLayout>
+	</div>
+</template>
```
Comment (danielroe) [inline playground/app.vue:7]: could we lint the project so we have consistent spacing? (looks like some files have tabs, some spaces)...

File: src/logger.ts [hunk 3]
```diff
@@ -0,0 +1,2 @@
+import consola from 'consola'
+export default consola.withScope('nuxt:yandex-metrika')
```
Comment (danielroe) [inline src/logger.ts:2]: we now have a handy new utility in kit which means we can also remove the consola dependency.
```suggestion
import { useLogger } from '@nuxt/kit'
export default useLogger('nuxt:yandex-metrika')
```

File: src/module.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	interface PublicRuntimeConfig {
+		[CONFIG_KEY]?: ModuleOptions
+	}
+}
+
+export default defineNuxtModule<ModuleOptions>({
+	meta: {
+		name: '@nuxtjs/yandex-metrika',
+		configKey: CONFIG_KEY
+	},
```
Comment (danielroe) [inline src/module.ts:40]: we could also add compatibility information here - e.g. does it require bridge, does it work with nuxt 2, etc.

File: src/runtime/nuxt2.plugin.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import { defineNuxtPlugin, useRuntimeConfig } from '#app'
+import { useRouter } from "#imports";
+
+export default defineNuxtPlugin(async (_) => {
+	const { id, metrikaUrl, ...metrikaOptions } = useRuntimeConfig().yandexMetrika
+
+	let ready = false
+	const basePath = (useRouter().options.base || '/').replace(/\/$/, '')
```
Comment (danielroe) [inline src/runtime/nuxt2.plugin.ts:8]: we can likely use runtime config to get this instead, here and in the nuxt3 version (perhaps we could unify them?):

```js
import { useRuntimeConfig } from '#imports'
// ...
const basePath = useRuntimeConfig().app.basePath
```

**Review by danielroe — COMMENTED** [review]
Thanks! Just a heads-up, we are also thinking of potentially merging this with some other modules, like the gtm module, for a unified approach to metrics/analytics.

cc: @pi0

File: playground/nuxt.config.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import { defineNuxtConfig } from 'nuxt3'
+import YandexMetrikaModule from '..'
+
+export default defineNuxtConfig({
+  dev: true,
+  render: {
+    resourceHints: false
+  },
```
Comment (danielroe) [inline playground/nuxt.config.ts:8]: I think we shouldn't need this.

```suggestion
```

**Review by danielroe — APPROVED** [review]
This looks great! ❤️

File: src/module.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    triggerEvent: false
+  },
+  setup (options: ModuleOptions, nuxt) {
+    // Don't include on dev mode
+    if (options.dev && process.env.NODE_ENV !== 'production') {
+      return
+    }
+
+    if (!options.id) {
+      logger.error('No id provided.')
```
Comment (IlyaSemenov) [inline src/module.ts:66]: Can we have a way to disable/not enable metrika in runtime without logging an error message? Such as for running e2e tests against the production build.

Comment (IlyaSemenov) [reply src/module.ts:68]: I'm not sure how is this resolved? Again, I'd like to be able to run `YANDEX_METRIKA_ID=xxx nuxt start` and have metrika enabled, then run `nuxt start` and have metrika script not loaded. I suppose with the current implementation that is going to be possible but will emit this error?

**Comment by dankerow** [comment]
> Thanks for your efforts in this nice PR @dankerow. We started working with @danielroe on a new `nuxt/scripts` module for Nuxt 3 support. Since there is a lot of shared logic between analytics modules and the rest of community integrations like GA and GTM. It would make more sense for maintenance to deprecate them and unify efforts.
> 
> New repo will be public in the upcoming days and of course, you are more than welcome to help on Yandex support in `nuxt/scripts`
> 
> (Keeping PR until making nuxt/scripts public)

Alright, no problem 👌

**Comment by melnikovdv** [comment]
It's 6 month since the last message. Could you merge it not to wait Google Analytics and other? 🙏🏻

**Comment by digitalcortex** [comment]
I have a problem with using dankerow's version of the module. After running `yarn add https://github.com/dankerow/yandex-metrika-module.git` my project refuses to run via `yarn dev` command.

The error message is following:
` ERROR  Cannot start nuxt:  Cannot find module '/Users/<...>/node_modules/@nuxtjs/yandex-metrika/dist/module.cjs'

  at createEsmNotFoundErr (node:internal/modules/cjs/loader:1016:15)
  at finalizeEsmResolution (node:internal/modules/cjs/loader:1009:15)
  at resolveExports (node:internal/modules/cjs/loader:529:14)
  at Module._findPath (node:internal/modules/cjs/loader:569:31)
  at Module._resolveFilename (node:internal/modules/cjs/loader:981:27)
  at Function.resolve (node:internal/modules/cjs/helpers:109:19)
  at Function._resolve [as resolve] (node_modules/jiti/dist/jiti.js:1:243302)
  at resolveModule (node_modules/nuxt/node_modules/@nuxt/kit/dist/index.mjs:240:29)
  at normalizeModule (node_modules/nuxt/node_modules/@nuxt/kit/dist/index.mjs:445:18)
  at installModule (node_modules/nuxt/node_modules/@nuxt/kit/dist/index.mjs:426:47)
  at initNuxt (node_modules/nuxt/dist/index.mjs:1787:13)
  at async load (node_modules/nuxi/dist/chunks/dev.mjs:6778:9)
  at async Object.invoke (node_modules/nuxi/dist/chunks/dev.mjs:6839:5)
  at async _main (node_modules/nuxi/dist/cli.mjs:50:20)`

Am I missing some installation instructions?

**Comment by krim** [comment]
I have the same problem as [digitalcortex](https://github.com/digitalcortex).
```js
 ERROR  Cannot start nuxt:  Cannot find module '/Users/kr/Sites/travel/frontend/node_modules/@nuxtjs/yandex-metrika/dist/module.cjs'                                                                                                                                             13:55:00

  at createEsmNotFoundErr (node:internal/modules/cjs/loader:960:15)
  at finalizeEsmResolution (node:internal/modules/cjs/loader:953:15)
  at resolveExports (node:internal/modules/cjs/loader:482:14)
  at Function.Module._findPath (node:internal/modules/cjs/loader:522:31)
  at Function.Module._resolveFilename (node:internal/modules/cjs/loader:919:27)
  at Function.resolve (node:internal/modules/cjs/helpers:108:19)
  at Function._resolve [as resolve] (node_modules/jiti/dist/jiti.js:1:243302)
  at resolveModule (node_modules/@nuxt/kit/dist/index.mjs:260:29)
  at normalizeModule (node_modules/@nuxt/kit/dist/index.mjs:429:18)
  at installModule (node_modules/@nuxt/kit/dist/index.mjs:415:47)
```

**Comment by digitalcortex** [comment]
> I have the same problem as [digitalcortex](https://github.com/digitalcortex).
> 
> ```js
>  ERROR  Cannot start nuxt:  Cannot find module '/<...>/node_modules/@nuxtjs/yandex-metrika/dist/module.cjs'                                                                                                                                             13:55:00
> 
>   at createEsmNotFoundErr (node:internal/modules/cjs/loader:960:15)
>   at finalizeEsmResolution (node:internal/modules/cjs/loader:953:15)
>   at resolveExports (node:internal/modules/cjs/loader:482:14)
>   at Function.Module._findPath (node:internal/modules/cjs/loader:522:31)
>   at Function.Module._resolveFilename (node:internal/modules/cjs/loader:919:27)
>   at Function.resolve (node:internal/modules/cjs/helpers:108:19)
>   at Function._resolve [as resolve] (node_modules/jiti/dist/jiti.js:1:243302)
>   at resolveModule (node_modules/@nuxt/kit/dist/index.mjs:260:29)
>   at normalizeModule (node_modules/@nuxt/kit/dist/index.mjs:429:18)
>   at installModule (node_modules/@nuxt/kit/dist/index.mjs:415:47)
> ```

I contacted the creator of the fork and he suggested to download the repository of the module and compile it. I ended up compiling it and putting the contents of the /dist folder into my project. After that I added this module the following way: 
`export default defineNuxtConfig({
     modules: [ './modules/yandex-metrika/module.cjs' ],
})`

Attaching the compiled contents of the yandex-metrika-module fork compatible with Nuxt 3:
[dankerow-yandex-metrika-module-build.zip](https://github.com/nuxt-community/yandex-metrika-module/files/10146215/dankerow-yandex-metrika-module-build.zip)

**Comment by digitalcortex** [comment]
I needed this module for my project so I created a precompiled NPM package of the dankerow's fork. It will be useful for those who want a working yandex metrika plugin without the need to compile it manually.

[https://www.npmjs.com/package/yandex-metrika-module-nuxt3](yandex-metrika-module-nuxt3)

**Pull request closed without merging**